File: Data Structures & Algorithms/subsets/submission-1.go
func subsets(nums []int) [][]int {
    n := len(nums)
	res := make([][]int, 0, 1<<n)
    subset := make([]int, 0, n)
    var dfs func(int)

    dfs = func(i int) {
        if i >= n {
            tmp := make([]int, len(subset))
            copy(tmp, subset)
            res = append(res, tmp)
            return
        }

        subset = append(subset, nums[i])
        dfs(i + 1)

        subset = subset[:len(subset)-1]
        dfs(i + 1)
    }

    dfs(0)
    return res

}

File: Data Structures & Algorithms/combination-target-sum/submission-1.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = []
        subset = []

        def dfs(i):
            subset_sum = sum(subset)
            if subset_sum == target:
                res.append(subset.copy())
                return
            elif subset_sum > target or i >= len(nums):
                return

            subset.append(nums[i])
            dfs(i)

            subset.pop()
            dfs(i + 1)
        
        dfs(0)
        return res